File: RPC/2020:06:13 Junio/editorial/radioprize/submissions/time_limit_exceeded/hamerly-exhaustive.py
#!/usr/bin/env python3

# try DFS from every node

import sys

def main():
    # read input, construct the tree
    ri = lambda: tuple(map(int, sys.stdin.readline().split()))
    n, = ri()
    taxes = ri()
    tree = [[] for _ in range(n)]
    for _ in range(n-1):
        u, v, w = ri()
        u, v = u - 1, v - 1
        tree[u].append((v, w))
        tree[v].append((u, w))

    for start in range(n):
        answer = 0
        stack = [(start, -1, 0)] # (current, predecessor, depth)
        while stack:
            u, p, depth = stack.pop()
            answer += (taxes[start] + taxes[u]) * depth
            for v, w in tree[u]:
                if v == p: # don't go back up the tree
                    continue
                stack.append((v, u, depth + w))

        print(answer)

if __name__ == '__main__':
    main()

File: RPC/2020:06:13 Junio/editorial/radioprize/submissions/time_limit_exceeded/hamerly-exhaustive.cpp
#include <iostream>
#include <vector>
#include <tuple>
#include <iomanip>

int main() {
    int n;
    std::cin >> n;
    std::vector<int> taxes(n);
    for (auto &t : taxes) std::cin >> t;
    std::vector<std::vector<std::pair<int, int>>> tree(n);
    for (int i = 0; i < n - 1; ++i) {
        int u, v, w;
        std::cin >> u >> v >> w;
        u--;
        v--;
        tree[u].push_back({v, w});
        tree[v].push_back({u, w});
    }

    for (int start = 0; start < n; ++start) {
        unsigned long long answer = 0;
        std::vector<std::tuple<int, int, unsigned long long>> stack = {{start, -1, 0}};

        while (stack.size()) {
            int u, p, v, w;
            unsigned long long depth;
            std::tie(u, p, depth) = stack.back();
            stack.pop_back();

            answer += (taxes[start] + taxes[u]) * depth;

            for (auto vw : tree[u]) {
                std::tie(v, w) = vw;
                if (v == p)
                    continue;
                stack.push_back({v, u, depth + w});
            }
        }

        std::cout << answer << "\n";
    }

    return 0;
}
